app/services/api_keys/track_usage_service.rb

```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+module ApiKeys
+  class TrackUsageService < BaseService
+    def call
+      ApiKey.find_each do |api_key|
```

Comment:
This is a bit ambitious!
You'll run over all the `ApiKey` event the one that are not used.
Is it better to loop over whats in cache?

Reply:
Indeed, that looks heavy but I still think that it’s acceptable option.

In order to do it another way (lookup in cache) we either need to know all keys we’re interested in which basically loading every api key to build cache keys. Or our cache store should have some pattern based lookup which will depend on particular cache store used which might vary for self hosted clients. Which is also solvable but that bring additional complexity to the solution.

Considering that all our clients have to use API I suppose that majority of key always in use. Maybe not each minute but each hour.

Speaking of speed improvements we can do 2 optimizations:
- Fetch API keys ids and iterate over it, looking for a cache key rather then loading whole active record object. That should reduce database load but it will have the same load on cache.
- Run this job less frequent (current each 5 minutes). E.g. each 1 hour.

What do you think?

Reply:
I understand!

I would say the best option is to delay it yes, but is it aligned with what product defined?